app/src/main/java/ch/hikemate/app/model/elevation/ElevationRepositoryCopernicus.kt

```diff
@@ -0,0 +1,399 @@
+ */
+data class ElevationRequest(
+    val coordinates: List<LatLong>,
+    val onSuccess: (List<Double>) -> Unit,
+    val onFailure: (Exception) -> Unit
+)
+
+/**
+ * A repository for the ElevationRepository. This class is responsible for making the network
+ * request
+ */
+class ElevationRepositoryCopernicus(
+    private val client: OkHttpClient,
+    private val repoDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val maxCacheSize: Int = DEFAULT_CACHE_MAX_SIZE
+) : ElevationRepository {
+
+  companion object {
+    // The base URL for the Elevation API
+    private const val BASE_URL = "https://www.elevation-api.eu/v1/elevation?pts="
+
+    // We need to wait some time before updating the cache to allow for multiple requests to be
+    // batched together
+    private const val WAITING_DELAY_BEFORE_SENDING_REQUEST_FOR_BATCHING = 500L
+
+    // Limit the cache size to avoid memory issues
+    // This has been tested with a cache size of 150000, on a Google Pixel 8 Pro, without any issues
+    const val DEFAULT_CACHE_MAX_SIZE = 100000
+
+    // The maximum number of failed requests before giving up
+    private const val MAX_FAILED_REQUESTS = 5
+
+    // Delay to add per failed request
+    private const val FAILED_REQUEST_DELAY = 1000L
+
+    // The max number of coordinates that can be sent in a single request
+    private const val MAX_COORDINATES_PER_REQUEST = 500
+
+    // The log tag
+    private val LOG_TAG = this::class.simpleName
+  }
+
+  // The number of failed requests made to the Elevation API since the last successful request
+  private var failedRequests = 0
+
+  // Cache for the elevation data, indexed by hike ID
+  private val cache = hashMapOf<LatLong, ElevationCacheEntry>()
+
+  // List of requests that are waiting for the cache to be updated
+  private val requests = mutableListOf<ElevationRequest>()
+
+  // A mutex to ensure that the cache and requests are accessed atomically
+  private val mutex: Mutex = Mutex()
+
+  // The job performing the cache update
+  private var cacheUpdateJob: Job? = null
+
+  /**
+   * Get the size of the cache
+   *
+   * @return The size of the cache
+   */
+  fun getCacheSize(): Int {
+    return cache.size
+  }
+
+  override fun getElevation(
+      coordinates: List<LatLong>,
+      onSuccess: (List<Double>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    if (coordinates.isEmpty()) {
+      onSuccess(emptyList())
+      return
+    }
+
+    CoroutineScope(repoDispatcher).launch {
+      mutex.withLock {
+        if (coordinates.minus(cache.keys).isEmpty()) {
+          // All coordinates are in the cache
+          onSuccess(
+              coordinates.map {
+                val value = cache[it]
+                if (value == null) {
+                  Log.e(LOG_TAG, "Coordinate is missing in cache")
+                  return@map 0.0
+                } else {
+                  value.elevation
+                }
+              })
```

Comment:
Suggestion: wouldn't it be more logical to first compute the elevation list, and if at least one of the elevation is `null`, call `onFailure` instead? It makes little sense to call `onSuccess` with a `null` elevation in my opinion, even for a single coordinate, and the default value of `0.0` is arbitrary.

Reply:
Just saw in `parseCacheOnDataRetrieval` that the same default value is being used, and it might take significant amounts of time to implement another approach.

I believe receiving a `null` value is relatively rare, rare enough that we keep this approach.

Reply:
Actually, `coordinates.minus(cache.keys).isEmpty()` ensures that the value must be in the cache if we reach those lines. This is just a fail-safe default (and to be type-compliant). Of course, we never store a null value in the cache